coderd/database/sqlxqueries/sqlx.go

```diff
@@ -0,0 +1,74 @@
+package sqlxqueries
+
+import (
+	"context"
+
+	"github.com/jmoiron/sqlx"
+
+	"golang.org/x/xerrors"
+)
+
+// constructQuery will return a SQL query by the given template name.
+// It will also return the arguments in order for the query based on the input
+// argument.
+func constructQuery(queryName string, argument any) (string, []any, error) {
+	// No argument was given, use an empty struct.
+	if argument == nil {
+		argument = struct{}{}
+	}
+
+	query, err := query(queryName, argument)
+	if err != nil {
+		return "", nil, xerrors.Errorf("get query: %w", err)
+	}
+
+	query, args, err := bindNamed(query, argument)
+	if err != nil {
+		return "", nil, xerrors.Errorf("bind named: %w", err)
+	}
+	return query, args, nil
+}
+
+// SelectContext runs the named query on the given database.
+// If the query returns no rows, an empty slice is returned.
+func SelectContext[RT any](ctx context.Context, q sqlx.QueryerContext, queryName string, argument any) ([]RT, error) {
+	var empty []RT
+	if q == nil {
+		return empty, xerrors.New("queryer is nil")
+	}
+
+	query, args, err := constructQuery(queryName, argument)
+	if err != nil {
+		return empty, xerrors.Errorf("get query: %w", err)
+	}
+
+	err = sqlx.SelectContext(ctx, q, &empty, query, args...)
+	if err != nil {
+		return empty, xerrors.Errorf("%s: %w", queryName, err)
+	}
+
+	return empty, nil
+}
+
+// GetContext runs the named query on the given database.
+// If the query returns no rows, sql.ErrNoRows is returned.
+func GetContext[RT any](ctx context.Context, q sqlx.QueryerContext, queryName string, argument interface{}) (RT, error) {
```

Comment:
If the only reason for these generics is so you don't need to instantiate the empty type before the function call then they should be rewritten to not use generics. Generics cause the function body to be duplicated for every type it's called with, which will probably be a lot for a function like this.

Reply:
It is also to match the return type so the caller does not need to do any casting.

Is that really an issue if it generates a new function for each type? We have to write 1 DB function for each SQL query, and a lot of queries reuse types. In this PR, I have 2 return types for all `WorkspaceBuild` types.

I don't think this is an issue. The total number of types will always be less than the total number of database functions.

Reply:
If you do this then you don't need to cast at all, so given that it's low effort and doesn't add complexity to avoid generics I think we should avoid them:

```go
var res []SomeType
err := SelectContext(ctx, q, "QueryName", arg, &res)
```